## Bài 1. Tạo Tensor cơ bản

In [1]:
import torch
a = torch.tensor(
    [[1, 2, 3], 
     [4, 5, 6]], 
    dtype=torch.int32
)
print(a)
print("Shape:", a.shape)
print("Số chiều (ndim):", a.ndim)
print("Kiểu dữ liệu (dtype):", a.dtype)

tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)
Shape: torch.Size([2, 3])
Số chiều (ndim): 2
Kiểu dữ liệu (dtype): torch.int32


Kích thước (shape) của Tensor a là 2x3 (2 hàng, 3 cột).

Số chiều (ndim) của Tensor a là 2.  

Kiểu dữ liệu (dtype) của Tensor a là torch.int32. 

## Bài 2. Tạo các Tensor thường dùng

In [2]:
import torch

a = torch.rand(3, 4)
b = torch.randn(3, 4)
c = torch.zeros(2, 3)
d = torch.ones(2, 3)
e = torch.eye(4)

print("a (rand):\n", a)
print("b (randn):\n", b)
print("c (zeros):\n", c)
print("d (ones):\n", d)
print("e (eye):\n", e)

# Tạo thêm một Tensor 5x5 chứa toàn giá trị 10
tensor_10 = torch.full((5, 5), 10)
print("Tensor 5x5 toàn 10:\n", tensor_10)

a (rand):
 tensor([[0.9456, 0.6022, 0.6902, 0.6035],
        [0.5310, 0.7365, 0.7666, 0.3189],
        [0.1737, 0.1505, 0.4976, 0.8651]])
b (randn):
 tensor([[ 0.3542, -0.6789,  0.7359,  0.8328],
        [ 1.2996,  0.9279,  1.0727,  0.2113],
        [-0.8829, -0.2566,  1.4927,  1.0032]])
c (zeros):
 tensor([[0., 0., 0.],
        [0., 0., 0.]])
d (ones):
 tensor([[1., 1., 1.],
        [1., 1., 1.]])
e (eye):
 tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])
Tensor 5x5 toàn 10:
 tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


Sự khác nhau giữa torch.rand() và torch.randn():

torch.rand(): Sinh ra một Tensor chứa các số ngẫu nhiên theo phân phối đều (uniform distribution) trong khoảng [0, 1).

torch.randn(): Sinh ra một Tensor chứa các số ngẫu nhiên theo phân phối chuẩn tắc (standard normal distribution) với giá trị trung bình (mean) bằng 0 và độ lệch chuẩn (variance) bằng 1.

Tạo Tensor 5x5 chứa toàn giá trị 10: Sử dụng hàm torch.full((5, 5), 10) như đã gắn trong đoạn code phía trên.

## Bài 3. Kiểm tra thông tin Tensor

In [3]:
import torch
a = torch.zeros(2, 3, 4)
print("Tensor:\n", a)
print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)

Tensor:
 tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions: 3
Data type: torch.float32


Ý nghĩa của torch.Size([2, 3, 4]): Tensor có tổng cộng 3 chiều (3D). Trong đó, chiều thứ nhất có 2 phần tử (có thể hiểu là 2 khối/ma trận), chiều thứ hai có 3 phần tử (3 hàng) và chiều thứ ba có 4 phần tử (4 cột). Tổng số phần tử bên trong là $2 \times 3 \times 4 = 24$ phần tử. 

## Bài 4. Slicing và Indexing

In [4]:
import torch
a = torch.randn(3, 4, 5)

print("\nTensor a[1]:")
print(a[1])

print("\nTensor a[1:, 2:4]:")
print(a[1:, 2:4])

print("\nHai phần tử cuối cùng trên chiều thứ ba:")
print(a[:, :, -2:])


Tensor a[1]:
tensor([[ 0.9179,  0.5586, -0.2078, -0.5281,  0.1812],
        [ 0.6230, -1.0346,  2.1794,  0.4630,  0.1539],
        [ 0.8667,  0.1751,  0.1609, -0.4153,  0.5178],
        [-0.3344,  1.3080, -0.4460,  0.2941, -1.8798]])

Tensor a[1:, 2:4]:
tensor([[[ 0.8667,  0.1751,  0.1609, -0.4153,  0.5178],
         [-0.3344,  1.3080, -0.4460,  0.2941, -1.8798]],

        [[-0.4122, -0.0355,  0.9828,  0.0342, -0.8981],
         [-1.5674, -0.6983, -0.5721,  0.6819, -0.8371]]])

Hai phần tử cuối cùng trên chiều thứ ba:
tensor([[[-0.7912,  1.1441],
         [ 0.4606, -1.0484],
         [-0.1305, -0.8786],
         [ 0.3291, -0.3502]],

        [[-0.5281,  0.1812],
         [ 0.4630,  0.1539],
         [-0.4153,  0.5178],
         [ 0.2941, -1.8798]],

        [[-0.5153, -2.3377],
         [ 1.3112, -0.8398],
         [ 0.0342, -0.8981],
         [ 0.6819, -0.8371]]])


Shape của các Tensor đã lấy:Shape của a[1] là torch.Size([4, 5]) (Lấy khối ma trận thứ 2 từ tensor 3D gốc).  

Shape của a[1:, 2:4] là torch.Size([2, 2, 5]) (Lấy từ khối 1 đến cuối, hàng từ 2 đến 3, và lấy hết 5 cột).  

Hai phần tử cuối cùng trên chiều thứ ba: Để truy xuất, ta dùng indexing a[:, :, -2:] như trong code, điều này có nghĩa là lấy tất cả các phần tử ở chiều 1 và 2, nhưng chiều thứ 3 chỉ lấy 2 cột cuối cùng.  

## Bài 5. Reshape và Flatten

In [5]:
import torch
a = torch.randn(3, 4)
print("Original shape:", a.shape)

a_reshape = a.reshape(3, 2, 2)
print("Reshape to 3x2x2 shape:", a_reshape.shape)

a_2x6 = a.reshape(2, 6)
print("Reshape to 2x6 shape:", a_2x6.shape)

Original shape: torch.Size([3, 4])
Reshape to 3x2x2 shape: torch.Size([3, 2, 2])
Reshape to 2x6 shape: torch.Size([2, 6])


Kiểm tra số phần tử trước và sau khi reshape: Số phần tử trước khi reshape là $3 \times 4 = 12$ phần tử. 

Sau khi reshape thành 3x2x2, số phần tử là $3 \times 2 \times 2 = 12$ phần tử. 

Phép reshape không làm thay đổi tổng số lượng phần tử cũng như dữ liệu gốc mà chỉ thay đổi cách biểu diễn kích thước (shape).  

Reshape thành 2x6: Được thực hiện thành công qua lệnh a.reshape(2, 6)

## Bài 6. Phép toán và nhân ma trận

In [6]:
import torch
a = torch.randn(2, 3)
b = torch.randn(2, 3)

print("a * b (Element-wise):")
print(a * b)

print("\na @ b.T (Matrix multiplication):")
print(a @ b.T)

a * b (Element-wise):
tensor([[-0.1785,  2.6405,  0.3415],
        [ 0.3061, -0.2925,  0.3835]])

a @ b.T (Matrix multiplication):
tensor([[ 2.8035, -1.4094],
        [ 0.6837,  0.3971]])


Phân biệt a * b và a @ b.T:a * b là phép nhân từng phần tử (element-wise multiplication). 

Cụ thể, các phần tử ở vị trí tương ứng của hai ma trận cùng kích thước sẽ nhân với nhau.  

a @ b.T là phép nhân ma trận đại số tuyến tính (matrix multiplication). Ma trận a ($2 \times 3$) nhân với ma trận chuyển vị của b là b.T ($3 \times 2$).  

Shape của kết quả a @ b.T: Theo nguyên tắc nhân ma trận $(2 \times 3) \times (3 \times 2)$, kết quả sẽ trả về một Tensor có kích thước (shape) là torch.Size([2, 2]).

## Bài 7. Các hàm thống kê trên Tensor

In [7]:
import torch
a = torch.randn(3, 4)

print("\nMean theo dim=1:")
print(torch.mean(a, dim=1))

print("\nStandard deviation theo dim=1:")
print(torch.std(a, dim=1))

print("\nCumulative sum theo dim=1:")
print(torch.cumsum(a, dim=1))


Mean theo dim=1:
tensor([-0.4933,  0.7557, -0.2668])

Standard deviation theo dim=1:
tensor([0.8770, 1.1568, 1.6211])

Cumulative sum theo dim=1:
tensor([[-0.3914,  0.1024, -1.5394, -1.9731],
        [ 2.3503,  2.3134,  3.1789,  3.0227],
        [ 0.8886,  2.1681,  0.0465, -1.0672]])


hận xét sự khác nhau giữa dim=0 và dim=1:Khi dùng dim=0, các hàm thống kê thực hiện tính toán dọc theo chiều của các hàng (tính cho từng cột một). 

Do a có kích thước $3 \times 4$, kết quả của mean/std sẽ là một tensor 1D có 4 phần tử.  

Khi thay bằng dim=1, các hàm thống kê sẽ thực hiện tính toán dọc theo chiều của các cột (tính cho từng hàng một). 

Kết quả lúc này sẽ trả về một tensor 1D có 3 phần tử tương ứng với 3 hàng. 

## Bài 8. Autograd cơ bản

In [8]:
import torch

# Câu 1: y = x^2
x = torch.tensor(3.6, requires_grad=True)
y = x * x
y.backward()
print("Với y = x^2 tại x = 3.6:")
print("x.grad =", x.grad.item()) 

# Câu 2: y = 3x^2 + 2x + 1
x2 = torch.tensor(3.6, requires_grad=True)
y2 = 3 * (x2 ** 2) + 2 * x2 + 1
y2.backward()
print("Với y = 3x^2 + 2x + 1 tại x = 3.6:")
print("x2.grad =", x2.grad.item())

Với y = x^2 tại x = 3.6:
x.grad = 7.199999809265137
Với y = 3x^2 + 2x + 1 tại x = 3.6:
x2.grad = 23.599998474121094


Tính thủ công $y = x^2$ tại $x = 3.6$: Đạo hàm $y' = 2x$. Thay $x = 3.6 \Rightarrow y' = 2 \times 3.6 = 7.2$. 

Kết quả này hoàn toàn khớp với giá trị x.grad do PyTorch tự động tính.  

Hàm $y = 3x^2 + 2x + 1$: Đạo hàm thủ công là $y' = 6x + 2$. Tại $x = 3.6 \Rightarrow y' = 6(3.6) + 2 = 21.6 + 2 = 23.6$. 

Code chạy cho ra x2.grad bằng 23.6, khớp với thực tế.  

## Bài 9. Gradient Descent tìm hệ số đa thức

In [9]:
import numpy as np
import torch

# Đa thức y = x^2 + 2x + 3
polynomial = np.poly1d([1, 2, 3])
N = 20
X = np.random.randn(N, 1) * 5
Y = polynomial(X)

XX = np.hstack([X*X, X, np.ones_like(X)])

# Khởi tạo trọng số ngẫu nhiên
w = torch.randn(3, 1, requires_grad=True)
x_tensor = torch.tensor(XX, dtype=torch.float32)
y_tensor = torch.tensor(Y, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)

for epoch in range(2000): # Tăng epoch để hội tụ tốt hơn
    optimizer.zero_grad()
    y_pred = x_tensor @ w
    mse = torch.mean(torch.square(y_tensor - y_pred))
    mse.backward()
    optimizer.step()

print("Learned coefficients:\n", w.detach().numpy())

Learned coefficients:
 [[1.0025992]
 [2.0000682]
 [3.0008364]]


So sánh hệ số học được với [1, 2, 3]: Sau quá trình huấn luyện, Tensor w sẽ cập nhật và dần tiệm cận đến các giá trị $\approx 1.0, \approx 2.0$ và $\approx 3.0$. Các sai số cực nhỏ còn lại là do learning rate và số lượng epoch hữu hạn.  

Giải thích vai trò của các hàm:optimizer.zero_grad(): Xóa sạch (reset) các giá trị đạo hàm (gradient) đã tính ở vòng lặp (epoch) trước đó để tránh hiện tượng gradient bị cộng dồn liên tục.  

mse.backward(): Chạy thuật toán Backpropagation (lan truyền ngược) tự động để tính toán đạo hàm (gradient) của hàm mất mát mse theo tất cả các Tensor có requires_grad=True (ở đây là biến w).  

optimizer.step(): Sử dụng gradient vừa tính được để tiến hành cập nhật/điều chỉnh giá trị của hệ số w theo đúng thuật toán tối ưu NAdam, giúp giảm dần sai số.  

## Bài 10. Giải hệ phương trình bằng Autograd

In [10]:
import random
import torch

A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)

EPOCHS = 2000
optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)

for _ in range(EPOCHS):
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2
    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4
    
    optimizer.zero_grad()
    sqerr.backward()
    optimizer.step()

print(f"A = {A.item():.4f}, B = {B.item():.4f}, C = {C.item():.4f}, D = {D.item():.4f}")
print("A + B =", (A + B).item())
print("C - D =", (C - D).item())
print("A + C =", (A + C).item())
print("B - D =", (B - D).item())

A = 4.7160, B = 4.2840, C = 3.2840, D = 2.2840
A + B = 8.999950408935547
C - D = 1.0000112056732178
A + C = 7.999967098236084
B - D = 1.9999940395355225


Kiểm tra nghiệm thu được: Sau quá trình chạy, khi in tổng/hiệu của các biến ngược trở lại vào 4 phương trình, kết quả sẽ xấp xỉ chính xác với vế phải ($9, 1, 8, 2$) với sai số vô cùng nhỏ.  

Nhận xét vì sao bài toán có thể có nhiều nghiệm: Ta có 4 phương trình: $A+B=9$ (1), $C-D=1$ (2), $A+C=8$ (3), $B-D=2$ (4). Nếu lấy (1) trừ đi (3), ta được $B-C=1$. Lại lấy (4) trừ (2) ta cũng được $(B-D) - (C-D) = B-C = 1$. 

Điều này cho thấy có một phương trình bị dư (phụ thuộc tuyến tính vào các phương trình kia). Hệ gồm 4 ẩn nhưng chỉ có 3 phương trình độc lập, do vậy hệ này vô định và có vô số nghiệm.

## Bài 11. Bài tập áp dụng (Tổng hợp)

In [11]:
import torch

print("--- Yêu cầu 1 ---")
X = torch.arange(24)
print("Reshape 4x6:\n", X.reshape(4, 6))
print("Reshape 2x12:\n", X.reshape(2, 12))
print("Reshape 2x3x4:\n", X.reshape(2, 3, 4))

print("\n--- Yêu cầu 2 ---")
A = torch.rand(3, 4)
B = torch.rand(3, 4)
print("A + B:\n", A + B)
print("A * B:\n", A * B)
print("A / B:\n", A / B)
print("Trung bình theo cột (dim=0):\n", torch.mean(A, dim=0))
print("Độ lệch chuẩn theo cột (dim=0):\n", torch.std(A, dim=0))

print("\n--- Yêu cầu 3 ---")
x3 = torch.tensor(2.0, requires_grad=True)
y3 = 2 * (x3 ** 2) + 5 * x3 + 3
y3.backward()
print("Autograd dy/dx tại x=2:", x3.grad.item())

print("\n--- Yêu cầu 4 ---")
import numpy as np

# y = 2x^2 - 3x + 5
polynomial_new = np.poly1d([2, -3, 5]) 
N = 20
X_new = np.random.randn(N, 1) * 5
Y_new = polynomial_new(X_new)

XX_new = np.hstack([X_new*X_new, X_new, np.ones_like(X_new)])

w_new = torch.randn(3, 1, requires_grad=True)
x_tensor_new = torch.tensor(XX_new, dtype=torch.float32)
y_tensor_new = torch.tensor(Y_new, dtype=torch.float32)

opt = torch.optim.NAdam([w_new], lr=0.05)

for epoch in range(2500):
    opt.zero_grad()
    y_pred = x_tensor_new @ w_new
    loss = torch.mean(torch.square(y_tensor_new - y_pred))
    loss.backward()
    opt.step()

print("Hệ số tìm được cho đa thức y = 2x^2 - 3x + 5:\n", w_new.detach().numpy())

--- Yêu cầu 1 ---
Reshape 4x6:
 tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])
Reshape 2x12:
 tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])
Reshape 2x3x4:
 tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])

--- Yêu cầu 2 ---
A + B:
 tensor([[0.9982, 1.0245, 1.1655, 1.4818],
        [1.7034, 0.7335, 0.5954, 0.7810],
        [0.8611, 1.3463, 1.0675, 1.3366]])
A * B:
 tensor([[0.1022, 0.2374, 0.1780, 0.5446],
        [0.7254, 0.0968, 0.0758, 0.1514],
        [0.0619, 0.3712, 0.1723, 0.4319]])
A / B:
 tensor([[7.6202, 1.8924, 5.4478, 0.8360],
        [0.9841, 3.2526, 2.2268, 1.1867],
        [9.8878, 2.4801, 4.3855, 0.6924]])
Trung bình theo cột (dim=0):
 tensor([0.8365, 0.7303, 0.7550, 0.5485])
Độ lệch chuẩn theo cột

Yêu cầu 3 tính $dy/dx$ cho hàm $y = 2x^2 + 5x + 3$. Đạo hàm thủ công: $y' = 4x + 5$. Tại $x = 2 \Rightarrow y' = 4(2) + 5 = 13$. 

Kết quả này khớp hoàn toàn với kết quả sinh ra từ lệnh .backward(). 

Yêu cầu 4, sau khi huấn luyện, mảng w_new sẽ trả về các giá trị xấp xỉ [2.0, -3.0, 5.0] tương ứng với 3 hệ số của đa thức mới. 